aimmo-game/simulation/action.py

```diff
@@ -1,145 +1,152 @@
-        if self.is_legal(world_map):
-            self.apply(world_map)
+    def process(self, game_state, other_actions=None):
+        if self.is_legal(game_state, other_actions):
+            self.apply(game_state, other_actions)
         else:
-            self.reject()
+            self.reject(game_state)
 
-    def is_legal(self, world_map):
+    def is_legal(self, game_state, other_actions=None):
         raise NotImplementedError('Abstract method')
 
-    def apply(self, world_map):
+    def apply(self, game_state, other_actions):
         raise NotImplementedError('Abstract method')
 
-    def reject(self):
+    def reject(self, game_state):
         raise NotImplementedError('Abstract method')
 
 
 class WaitAction(Action):
-    def __init__(self, avatar):
-        super(WaitAction, self).__init__(avatar)
+    def __init__(self, avatar, source):
+        Action.__init__(self, avatar, source)
 
-    def is_legal(self, world_map):
+    def is_legal(self, game_state, other_actions=None):
         return True
 
-    def apply(self, world_map):
-        self.avatar.clear_action()
+    def apply(self, game_state, other_actions):
+        pass
 
 
 class MoveAction(Action):
-    def __init__(self, avatar, direction):
+    def __init__(self, avatar, source, direction):
         # Untrusted data!
-        self.direction = Direction(**direction)
-        super(MoveAction, self).__init__(avatar)
+        Action.__init__(self, avatar, source, direction=Direction.copy(direction))
 
-    def is_legal(self, world_map):
-        return world_map.can_move_to(self.target_location)
+    def process(self, game_state, other_actions=None):
+        self.chain(game_state, other_actions, {self.source})
 
-    def process(self, world_map):
-        self.chain(world_map, {self.avatar.location})
+    def chain(self, game_state, other_actions, visited):
+        if not self.is_legal(game_state, other_actions):
+            return self.reject(game_state)
 
-    def apply(self, world_map):
-        event = MovedEvent(self.avatar.location, self.target_location)
-        self.avatar.add_event(event)
+        # Detect cycles
+        if self.target in visited:
+            return self.reject(game_state)
 
-        world_map.get_cell(self.avatar.location).avatar = None
-        self.avatar.location = self.target_location
-        world_map.get_cell(self.target_location).avatar = self.avatar
-        self.avatar.clear_action()
+        if not game_state.cell_occupied(self.target):
+            return self.apply(game_state, other_actions)
 
-        new_cell = world_map.get_cell(self.target_location)
-        if new_cell.pickup:
-            # TODO:  extract pickup logic into pickup when adding multiple types
-            self.avatar.health = min(10, self.avatar.health + new_cell.pickup.health_restored)
-            new_cell.pickup = None
+        next_avatar = game_state.avatar_at(self.target)
+        next_action = other_actions.by_avatar(next_avatar)
```

Comment:
Fails if `other_actions is None` (I can't see any reason to assume not None here).